cpp/src/arrow/type.h

```diff
@@ -1196,6 +1196,49 @@ class ARROW_EXPORT DenseUnionType : public UnionType {
   std::string name() const override { return "dense_union"; }
 };
 
+/// \brief Base class for types using a special-encoding (e.g. run-end-encoding)
+///
+/// This allows users to check if arrays are logically compatible even when the
+/// physical encoding of the types differs.
+class ARROW_EXPORT EncodedType : public NestedType {
+ public:
+  EncodedType(Type::type id, std::shared_ptr<DataType> encoded_type)
+      : NestedType(id), encoded_type_{std::move(encoded_type)} {}
+
+  const std::shared_ptr<DataType>& encoded_type() const { return encoded_type_; }
```

Comment:
Note that `DictionaryType` uses `value_type` not `encoded_type`. It would also match the child name which is "values".

Reply:
Renamed to `value_type` everywhere. And `values_array()` is now just `values()`.